Add dedicated integration with Product Bundles

tomalec opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Add dedicated integration with Product Bundles,
Use `->get_bundle_price( 'min' );` instead of `->get_price()` for bundles.

Fixes https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/219 the simple way.

### Checks:
<!-- Mark completed items with an [x] -->
* [x] Does your code follow the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Screenshots:

![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/0f16bc58-89cc-4464-983a-e06eeacd174f)


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Install and activate [the Product Bundles (PB) extension](https://woocommerce.com/products/product-bundles/)
2. Add a new bundle product
   ![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/3331d43f-4a9f-421d-8277-a13ae5128339)
3. Add bundled products that are priced individually 
   ![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/17435/afa6a5dd-5311-46a6-85ff-e6d9185568c9)
4. Publish the product bundle
5. Open the site through https://tagassistant.google.com/
6. Go to the product bundle page
7. Check that `view_item` event is tracked with the correct bundle price
8. Add the bundle to cart
9. Check that `add_to_cart` event is tracked with the correct bundle price
10. Uninstall the PB extension and smoke test for regression.



### Additional details:

1. I didn't add E2E tests, as PB is a paid extension, and public contributors to this repo would not be able to use it to run E2Es locally. We can try mocking PB interface, but then it's not really E2E. I'm open to better ideas :) 

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Add - WooCommerce Product Bundles integration.

tomalec committed:
Add dedicated integration with Product Bundles,

Use `->get_bundle_price( 'min' );` instead of `->get_price()` for bundles.

Fixes https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/219 the simples way.

martynmjones commented on includes/class-wc-abstract-google-analytics-js.php:
```diff
@@ -230,6 +230,13 @@ public function get_formatted_product( WC_Product $product, $variation_id = 0, $
+		// Integration with Product Bundles.
+		// Get the minimum price, as `get_price` may return 0 if the product is a bundle and the price is potentially a range.
+		// Even a range containing a single value.
+		if ( class_exists( '\WC_Product_Bundle' ) && $product instanceof WC_Product_Bundle && is_callable( [ $product, 'get_bundle_price' ] ) ) {
```
Had you considered simplifying this to:

```suggestion
		if ( $product->is_type( 'bundle' ) && is_callable( [ $product, 'get_bundle_price' ] ) ) {
```

Granted another extension may be registering it's own product type called `bundle` but in that scenario the class name may also match and as we have the second check to confirm the `get_bundle_price` method exists I think it would be suitable.

martynmjones reviewed (commented):
Hey @tomalec, thanks for working on the fix for bundle pricing!

Events tracked via the classic integration all look to be using the correct values for bundle prices.

There is still an issue with the Block `remove_from_cart` event. When we're [building the product object in JS](https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/55ef826168fd8f3f7907b53d15f0e8f106e14c1a/assets/js/src/utils/index.js#L12) we're making use of `product.prices.price` and when a product is removed from cart on a block powered page that value will be the base price:

<img width="1022" alt="Screenshot 2024-06-18 at 13 42 53" src="https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/40762232/b970c730-e3b6-4877-9223-37b4f0288569">

We could instead rely on `totals.line_total` for the price if it is available which should give us consistent pricing across different integrations.

<img width="991" alt="Screenshot 2024-06-18 at 13 50 08" src="https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/40762232/51c66ece-1562-4262-abd7-622ab54f7bf1">

---

> I didn't add E2E tests, as PB is a paid extension, and public contributors to this repo would not be able to use it to run E2Es locally. We can try mocking PB interface, but then it's not really E2E. I'm open to better ideas :)

I don't have much else to suggest here I'm afraid. We could add a unit test in PHP for the bundled price in `get_formatted_product` which would be a lot easier to mock although that wouldn't cover Blocks.

tomalec committed:
Simplify bundle price identification

https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/439#discussion_r1644397118

Co-authored-by: martynmjones <40762232+martynmjones@users.noreply.github.com>

tomalec committed:
Add support for bundled product price in Block handlers

Address https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/439#pullrequestreview-2125408349

tomalec commented:
Thanks, @martynmjones, for catching that!
I applied your suggestions. Speaking of tests, as there are none for `get_formatted_product` so far, WDYT of adding them as a follow-up in the next porter or cooldown cycle?

martynmjones reviewed (approved):
Thanks for the changes @tomalec!

Pricing is consistent across different events and integrations now so LGTM ✅ 

> WDYT of adding them as a follow-up in the next porter or cooldown cycle?

Agreed! A follow-up for adding tests sounds good.

tomalec commented:
Thanks :) Created the follow-up issue to track tests https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/442
